ngraph/core/src/op/proposal.cpp

```diff
@@ -75,6 +75,14 @@ void op::v0::Proposal::validate_and_infer_types()
             image_shape_shape[0],
             ").");
 
+        NODE_VALIDATION_CHECK(this,
```

Comment:
The high-level check `if (class_probs_pshape.is_static() && class_bbox_deltas_pshape.is_static() && image_shape_pshape.is_static())` must be split into 3 separate `if`s where the rank of the input is checked. Then you should use the partial shape instead of a shape and perform the checks for a particular dimension.

When you compute the batch dimension for the output you can use `Interval`s to calculate the output shape. So even if the `class_probs_pshape[0]` is dynamic you still can multiply it with a constant m_attrs.post_nms_topn and set it as a output batch dimension.

Reply:
Thanks for the input, and the help offline. Changed those checks to verify inputs one by one, also added (what I think is) better shape inference for the output.